File: DialedInCoffee/Models/Milk.swift
import Foundation

public enum MilkKind: String, Codable, CaseIterable, Hashable, Sendable, Identifiable {
    case devondaleFullCreamUHT
    case fullCreamFresh
    case skim
    case lactoseFree
    case oat
    case soy
    case almond

    public var id: String { rawValue }

    public var displayName: String {
        switch self {
        case .devondaleFullCreamUHT: return "Devondale Full Cream UHT"
        case .fullCreamFresh: return "Full Cream (Fresh)"
        case .skim: return "Skim"
        case .lactoseFree: return "Lactose-Free"
        case .oat: return "Oat"
        case .soy: return "Soy"
        case .almond: return "Almond"
        }
    }
}

public struct Milk: Codable, Hashable, Sendable {
    public var kind: MilkKind
    public var sweetness: Double       // 0...1
    public var bodyWeight: Double      // 0...1 (perceived body / fat)
    public var foamability: Double     // 0...1
    public var perceivedBitterness: Double // 0...1, how much it amplifies bitterness

    public init(
        kind: MilkKind,
        sweetness: Double,
        bodyWeight: Double,
        foamability: Double,
        perceivedBitterness: Double
    ) {
        self.kind = kind
        self.sweetness = sweetness
        self.bodyWeight = bodyWeight
        self.foamability = foamability
        self.perceivedBitterness = perceivedBitterness
    }

    public static let devondaleFullCreamUHT = Milk(
        kind: .devondaleFullCreamUHT,
        sweetness: 0.65,
        bodyWeight: 0.75,
        foamability: 0.7,
        perceivedBitterness: 0.35
    )

    public static let fullCreamFresh = Milk(
        kind: .fullCreamFresh,
        sweetness: 0.7,
        bodyWeight: 0.8,
        foamability: 0.85,
        perceivedBitterness: 0.3
    )

    public static let skim = Milk(
        kind: .skim,
        sweetness: 0.55,
        bodyWeight: 0.35,
        foamability: 0.9,
        perceivedBitterness: 0.55
    )

    public static let lactoseFree = Milk(
        kind: .lactoseFree,
        sweetness: 0.8,
        bodyWeight: 0.7,
        foamability: 0.75,
        perceivedBitterness: 0.3
    )

    public static let oat = Milk(
        kind: .oat,
        sweetness: 0.6,
        bodyWeight: 0.65,
        foamability: 0.6,
        perceivedBitterness: 0.5
    )

    public static let soy = Milk(
        kind: .soy,
        sweetness: 0.5,
        bodyWeight: 0.55,
        foamability: 0.55,
        perceivedBitterness: 0.6
    )

    public static let almond = Milk(
        kind: .almond,
        sweetness: 0.4,
        bodyWeight: 0.35,
        foamability: 0.4,
        perceivedBitterness: 0.7
    )

    public static func canonical(for kind: MilkKind) -> Milk {
        switch kind {
        case .devondaleFullCreamUHT: return .devondaleFullCreamUHT
        case .fullCreamFresh: return .fullCreamFresh
        case .skim: return .skim
        case .lactoseFree: return .lactoseFree
        case .oat: return .oat
        case .soy: return .soy
        case .almond: return .almond
        }
    }

    public static let allCanonical: [Milk] = MilkKind.allCases.map(Milk.canonical(for:))
}

File: android/core/src/main/kotlin/com/cawfee/domain/model/Milk.kt
package com.cawfee.domain.model

/** Milk properties that inform rule evaluation. Ported from Milk.swift. */
data class Milk(
    val kind: MilkKind,
    val sweetness: Double,            // 0..1
    val bodyWeight: Double,           // 0..1 (perceived body / fat)
    val foamability: Double,          // 0..1
    val perceivedBitterness: Double,  // 0..1, how much it amplifies bitterness
) {
    companion object {
        val DEVONDALE_FULL_CREAM_UHT = Milk(MilkKind.DEVONDALE_FULL_CREAM_UHT, 0.65, 0.75, 0.7, 0.35)
        val FULL_CREAM_FRESH = Milk(MilkKind.FULL_CREAM_FRESH, 0.7, 0.8, 0.85, 0.3)
        val SKIM = Milk(MilkKind.SKIM, 0.55, 0.35, 0.9, 0.55)
        val LACTOSE_FREE = Milk(MilkKind.LACTOSE_FREE, 0.8, 0.7, 0.75, 0.3)
        val OAT = Milk(MilkKind.OAT, 0.6, 0.65, 0.6, 0.5)
        val SOY = Milk(MilkKind.SOY, 0.5, 0.55, 0.55, 0.6)
        val ALMOND = Milk(MilkKind.ALMOND, 0.4, 0.35, 0.4, 0.7)

        fun canonical(kind: MilkKind): Milk = when (kind) {
            MilkKind.DEVONDALE_FULL_CREAM_UHT -> DEVONDALE_FULL_CREAM_UHT
            MilkKind.FULL_CREAM_FRESH -> FULL_CREAM_FRESH
            MilkKind.SKIM -> SKIM
            MilkKind.LACTOSE_FREE -> LACTOSE_FREE
            MilkKind.OAT -> OAT
            MilkKind.SOY -> SOY
            MilkKind.ALMOND -> ALMOND
        }

        val allCanonical: List<Milk> = MilkKind.entries.map { canonical(it) }
    }
}
